src/ssr-template-helper.js

```diff
@@ -0,0 +1,136 @@
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {Capability} from './service/viewer-impl';
+import {isArray} from './types';
+import {map} from './utils/object';
+
+/** The attributes we allow to be sent to the viewer. */
+const ATTRS_TO_SEND_TO_VIEWER = {
+  'amp-list': ['src', 'single-item', 'max-items'],
+  'amp-form': ['action-xhr'],
+};
+
+/**
+ * Helper, that manages the proxying of template rendering to the viewer.
+ */
+export class SsrTemplateHelper {
+
+  /**
+   * @param {string} sourceComponent
+   * @param {!./service/viewer-impl.Viewer} viewer
+   * @param {!./service/template-impl.Templates} templates
+   */
+  constructor(sourceComponent, viewer, templates) {
+
+    /** @private @const */
+    this.viewer_ = viewer;
+
+    /** @private @const */
+    this.templates_ = templates;
+
+    /** @private @const {!XMLSerializer} */
+    this.xmls_ = new XMLSerializer();
+
+    /** @private @const */
+    this.sourceComponent_ = sourceComponent;
+  }
+
+  /**
+   * Whether the viewer can render templates.
+   * @return {boolean}
+   */
+  isSupported() {
+    return this.viewer_.canRenderTemplates();
+  }
+
+  /**
+   * Proxies xhr and template rendering to the viewer and renders
+   * the response.
+   * @param {!Element} element
+   *     viewer to determine the component responsible for proxying the
+   *     request.
+   * @param {?function():Promise<?>} onSuccess
+   * @param {?function():Promise<?>} onFailure
+   * return {!Promise}
+   */
+  fetchAndRenderTemplate(
+    element,
+    onSuccess,
+    onFailure) {
+    const inputsAsJson = this.getElementInputsAsJson_(element);
+    const elementAttrsAsJson =
+        this.getElementAttributesAsJson_(element);
+    elementAttrsAsJson.inputData = inputsAsJson;
+    const template = this.templates_.maybeFindTemplate(element);
+    let mustacheTemplate;
+    if (template) {
+      // The document fragment can't be used in the message channel API thus
+      // serializeToString for a string representation of the dom tree.
+      mustacheTemplate = this.xmls_.serializeToString(
+          this.templates_.findTemplate(element));
+    }
+    return this.viewer_.sendMessageAwaitResponse(
+        Capability.VIEWER_RENDER_TEMPLATE,
+        {
+          data: elementAttrsAsJson,
+          mustacheTemplate,
+          'sourceAmpComponent': this.sourceComponent_,
+        })
+        .then(resp => {
+          return onSuccess ? onSuccess(resp) : resp;
+        }, errorResponseJson => {
+          return onFailure ?
+            onFailure(errorResponseJson) : errorResponseJson;
+        });
+  }
+
+  /**
+   * Returns the element's contained inputs and values in json format.
+   * @param {!HtmlElement} element
```

Comment:
nit: I think the rest of the project uses `{!Element}`

Reply:
done.